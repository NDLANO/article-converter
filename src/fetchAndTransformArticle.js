import cheerio from 'cheerio';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';

export async function transformArticle(
  article,
  lang,
  accessToken,
  options = {}
) {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;
  const { html, embedMetaData } = articleContent
    ? await transform(
        articleContent,
        lang,
        accessToken,
        article.visualElement,
        options
      )
    : {};
  const copyText = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang
  );

  return {
    ...article,
    content: html || '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction
      ? article.introduction.introduction
      : undefined,
    metaDescription: article.metaDescription
      ? article.metaDescription.metaDescription
      : '',
  };
}

export default async function fetchAndTransformArticle(
  articleId,
  lang,
  accessToken,
  options = {}
) {
  const article = await fetchArticle(articleId, accessToken, lang);
  const transformedArticle = await transformArticle(
    article,
    lang,
    accessToken,
    options
  );
  return transformedArticle;
}
