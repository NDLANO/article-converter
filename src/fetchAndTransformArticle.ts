import cheerio from 'cheerio';
import { IArticleV2 } from '@ndla/types-article-api';
import { fetchArticle } from './api/articleApi';
import { transform } from './transformers';
import config from './config';
import { getCopyString } from './plugins/pluginHelpers';
import { LocaleType, ResponseHeaders, TransformedArticle, TransformOptions } from './interfaces';

export async function transformArticle(
  article: IArticleV2,
  headers: ResponseHeaders,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  options: TransformOptions = {},
): Promise<TransformedArticle> {
  const articleContent = article.content.content
    ? cheerio.load(article.content.content, {
        recognizeSelfClosing: true,
      })
    : undefined;

  const { html, embedMetaData, responseHeaders } = articleContent
    ? await transform(
        articleContent,
        headers,
        lang,
        accessToken,
        feideToken,
        article.visualElement,
        options,
      )
    : { html: undefined, embedMetaData: undefined, responseHeaders: undefined };

  const htmlString: string = html ?? '';

  const copyText: string = getCopyString(
    article.title.title,
    config.ndlaFrontendDomain,
    options.path,
    article.copyright,
    lang,
  );
  const hasContent = article.articleType === 'standard' || cheerio.load(htmlString).text() !== '';

  return {
    ...article,
    content: hasContent ? htmlString : '',
    metaData: { ...embedMetaData, copyText } || '',
    title: article.title.title || '',
    tags: article.tags ? article.tags.tags : [],
    introduction: article.introduction ? article.introduction.introduction : undefined,
    metaDescription: article.metaDescription ? article.metaDescription.metaDescription : '',
    headerData: responseHeaders ?? {},
  };
}

export default async function fetchAndTransformArticle(
  articleId: string,
  lang: LocaleType,
  accessToken: string,
  feideToken: string,
  options = {},
): Promise<TransformedArticle> {
  const [article, headers] = await fetchArticle(articleId, accessToken, feideToken, lang);
  return await transformArticle(article, headers, lang, accessToken, feideToken, options);
}
