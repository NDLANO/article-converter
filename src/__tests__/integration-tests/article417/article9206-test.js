/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import nock from 'nock';
import bunyan from 'bunyan';
import { prettify } from '../../testHelpers';
import article9206 from './article9206';
import articleResource1128 from './articleResource1128';
import articleResource9202 from './articleResource9202';
import fetchAndTransformArticle from '../../../fetchAndTransformArticle';
import log from '../../../utils/logger';

const resources = {
  '1128': articleResource1128,
  '9202': articleResource9202,
  '1129': articleResource1128,
};

test('app/fetchAndTransformArticle 9206', async () => {
  nock('http://ndla-api')
    .get('/article-api/v2/articles/9206?language=nb&fallback=true')
    .reply(200, article9206);
  ['1128', '9202', '1129'].forEach(id => {
    nock('http://ndla-api')
      .get(`/article-api/v2/articles/${id}?language=nb&fallback=true`)
      .reply(200, article9206);
    nock('http://ndla-api')
      .get(
        `/taxonomy/v1/resources?contentURI=urn:article:${id}&language=nb`
      )
      .reply(200, resources[id]);
  });

  log.level(bunyan.FATAL + 1); // temporarily disable logging

  const transformed = await fetchAndTransformArticle(
    '9206',
    'nb',
    'some_token'
  );
  const { content, ...rest } = transformed;

  log.level(bunyan.INFO);

  expect(rest).toMatchSnapshot();
  expect(prettify(content)).toMatchSnapshot();
});
