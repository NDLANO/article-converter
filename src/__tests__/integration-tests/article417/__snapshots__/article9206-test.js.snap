// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/fetchAndTransformArticle 9206 1`] = `
Object {
  "articleType": "standard",
  "copyright": Object {
    "creators": Array [
      Object {
        "name": "Johannes Leiknes Nag",
        "type": "Writer",
      },
    ],
    "license": Object {
      "description": "Creative Commons Attribution-ShareAlike 2.0 Generic",
      "license": "CC-BY-SA-4.0",
      "url": "https://creativecommons.org/licenses/by-sa/2.0/",
    },
    "origin": "",
    "processors": Array [],
    "rightsholders": Array [],
  },
  "created": "2016-03-22T08:46:42Z",
  "headerData": Object {},
  "id": 9206,
  "introduction": "Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.",
  "metaData": Object {
    "copyText": "Nag, J. L. (2017, 7. mars). Skripting gjør sidene mer interaktive. NDLA. https://test.ndla.no/article/9206",
  },
  "metaDescription": "Vi bruker skript for å tilføre funksjonalitet og interaktivitet på nettsider utover det vi kan lage med markeringsspråk og stilark. Det kan være alt fra små justeringer i design til hele spillkonsept.",
  "oldNdlaUrl": "//red.ndla.no/node/160589",
  "published": "2017-03-07T17:16:47Z",
  "requiredLibraries": Array [],
  "revision": 1,
  "supportedLanguages": Array [
    "nb",
    "nn",
  ],
  "tags": Array [
    "design",
    "koding",
    "nettside",
    "programmering",
    "tagger",
  ],
  "title": "Skripting gjør sidene mer interaktive",
  "updated": "2017-03-07T17:16:47Z",
  "updatedBy": "content-import-client",
}
`;

exports[`app/fetchAndTransformArticle 9206 2`] = `
"<section>
  <figure class=\\"c-figure\\">
    <img
      alt=\\"En feil oppstod ved lasting av en ekstern ressurs.\\"
      src=\\"data:image/svg+xml;charset=UTF-8,%3Csvg fill='%238A8888' height='400' viewBox='0 0 24 12' width='100%25' xmlns='http://www.w3.org/2000/svg' style='background-color: %23EFF0F2'%3E%3Cpath d='M0 0h24v24H0V0z' fill='none'/%3E%3Cpath transform='scale(0.3) translate(28, 8.5)' d='M11 15h2v2h-2zm0-8h2v6h-2zm.99-5C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2zM12 20c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z'/%3E%3C/svg%3E\\"
    />
    <figcaption>En feil oppstod ved lasting av en ekstern ressurs.</figcaption>
  </figure>
</section>
<section>
  <p>
    Vil du lære deg å programmere? Da må du først velge et programmeringsspråk.
    Jobber du med nettsider, er det naturlig å lære seg JavaScript. Det er flere
    måter å komme i gang på. Du kan for eksempel følge et nettbasert kurs som
    det vi ser i videoen over, eller opplegget hos
    <a
      href=\\"https://nb.khanacademy.org/computing/computer-programming\\"
      rel=\\"noopener noreferrer\\"
      target=\\"_blank\\"
      title=\\"Khan Academy\\"
      >Khan Academy</a
    >. En annen tilnærming er å finne eksempelkoder som man tar i bruk på egne
    websider – vi kan kalle det problembasert tilnærming. Mange vil oppleve det
    svært nyttig.
  </p>
  <blockquote>
    «Everyone should learn how to program a computer because it teaches you how
    to think.» — Steve Jobs
  </blockquote>
  <p>
    Det finnes mange kodespråk. Noen av dem bygger kode med byggeklosser i
    stedet for ord – slik som
    <a
      href=\\"https://scratch.mit.edu/projects/editor/?tip_bar=getStarted\\"
      rel=\\"noopener noreferrer\\"
      target=\\"_blank\\"
      title=\\"Scratch på norsk\\"
      >Scratch på norsk</a
    >.
  </p>
  <style data-emotion=\\"css 8s0se0-ContainerDiv\\">
    .css-8s0se0-ContainerDiv {
      position: relative;
    }
    .css-8s0se0-ContainerDiv:hover button {
      cursor: pointer;
      opacity: 0.5;
    }
  </style>
  <div
    data-header-copy-container=\\"true\\"
    data-title=\\"Endre innhold\\"
    class=\\"css-8s0se0-ContainerDiv e18czas70\\"
  >
    <div>
      <style data-emotion=\\"css kgto6a-IconButton\\">
        .css-kgto6a-IconButton {
          position: absolute;
          left: -3em;
          top: 0.1em;
          background: none;
          border: 0;
          z-index: 1;
          -webkit-transition: 0.2s;
          transition: 0.2s;
          opacity: 0;
        }
        .css-kgto6a-IconButton svg {
          width: 30px;
          height: 30px;
        }</style
      ><button
        data-title=\\"Endre-innhold\\"
        class=\\"css-kgto6a-IconButton e18czas71\\"
      >
        <div
          data-tooltip=\\"true\\"
          data-tooltip-label=\\"Kopier lenke til overskriften\\"
          data-tooltip-from-article-converter=\\"true\\"
        >
          <style data-emotion=\\"css qyvxt5-StyledTooltip\\">
            .css-qyvxt5-StyledTooltip {
              color: #fff;
              z-index: 99999;
              border: 0;
              background: #444;
              border-radius: 4px;
              padding: 6.5px 13px;
              font-family: \\"Source Sans Pro\\", Helvetica, Arial, STKaiti,
                \\"华文楷体\\", KaiTi, SimKai, \\"楷体\\", KaiU, DFKai-SB, \\"標楷體\\",
                SongTi, \\"宋体\\", sans-serif;
              font-size: 14;
              line-height: 1.2;
              font-weight: 400;
              text-align: center;
              white-space: normal;
              max-width: calc(100vw - #{26px});
            }</style
          ><span
            data-tooltip-children=\\"true\\"
            data-state=\\"tooltip-hidden\\"
            data-reach-tooltip-trigger=\\"\\"
            ><svg
              fill=\\"currentColor\\"
              aria-hidden=\\"true\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align: middle\\"
            >
              <g>
                <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
                <path
                  d=\\"M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z\\"
                ></path>
              </g></svg
          ></span>
        </div>
      </button>
      <h2 id=\\"Endre-innhold\\" tabindex=\\"0\\">Endre innhold</h2>
    </div>
  </div>
  <p>
    Vi kan ha behov for å endre innhold på nettsidene våre basert på hva
    brukeren gjør. For enkelhets skyld viser vi her skript direkte i HTML-koden:
  </p>
  <pre><code>&lt;h1 <strong>onclick</strong>=\\"innerHTML='Lengre og mer utførlig overskrift'\\"&gt;Kort overskrift&lt;/h1&gt;<br></code></pre>
  <p>
    Hvis man klikker på overskriften «Kort overskrift», blir den byttet ut med
    en annen overskrift fordi skriptet endrer verdien innerHTML på H1-taggen.
  </p>
  <p>
    Vanligvis samles skriptene i et eget dokument, på samme måte som stilark. Vi
    kan da lage funksjonalitet som kan gjenbrukes i alle dokumenter, f.eks.
    innholdsfortegnelser:
  </p>
  <pre><code>var overskrifter = document.getElementByTagName(\\"h1\\");<br></code></pre>
  <p>
    Her vil alle overskrifter fanges opp og lagres, slik at de senere kan vises
    fram et annet sted – eller på forespørsel.
  </p>
  <p>
    I eksempelet over er <em>overskrifter</em> en <strong>variabel</strong> og
    <em>getElementByTagName</em> en <strong>funksjon</strong>. En funksjon
    beregner eller gjør noe, mens en variabel lagrer eller husker for oss.
  </p>
  <style data-emotion=\\"css 8s0se0-ContainerDiv\\">
    .css-8s0se0-ContainerDiv {
      position: relative;
    }
    .css-8s0se0-ContainerDiv:hover button {
      cursor: pointer;
      opacity: 0.5;
    }
  </style>
  <div
    data-header-copy-container=\\"true\\"
    data-title=\\"Endre stiler\\"
    class=\\"css-8s0se0-ContainerDiv e18czas70\\"
  >
    <div>
      <style data-emotion=\\"css kgto6a-IconButton\\">
        .css-kgto6a-IconButton {
          position: absolute;
          left: -3em;
          top: 0.1em;
          background: none;
          border: 0;
          z-index: 1;
          -webkit-transition: 0.2s;
          transition: 0.2s;
          opacity: 0;
        }
        .css-kgto6a-IconButton svg {
          width: 30px;
          height: 30px;
        }</style
      ><button
        data-title=\\"Endre-stiler\\"
        class=\\"css-kgto6a-IconButton e18czas71\\"
      >
        <div
          data-tooltip=\\"true\\"
          data-tooltip-label=\\"Kopier lenke til overskriften\\"
          data-tooltip-from-article-converter=\\"true\\"
        >
          <style data-emotion=\\"css qyvxt5-StyledTooltip\\">
            .css-qyvxt5-StyledTooltip {
              color: #fff;
              z-index: 99999;
              border: 0;
              background: #444;
              border-radius: 4px;
              padding: 6.5px 13px;
              font-family: \\"Source Sans Pro\\", Helvetica, Arial, STKaiti,
                \\"华文楷体\\", KaiTi, SimKai, \\"楷体\\", KaiU, DFKai-SB, \\"標楷體\\",
                SongTi, \\"宋体\\", sans-serif;
              font-size: 14;
              line-height: 1.2;
              font-weight: 400;
              text-align: center;
              white-space: normal;
              max-width: calc(100vw - #{26px});
            }</style
          ><span
            data-tooltip-children=\\"true\\"
            data-state=\\"tooltip-hidden\\"
            data-reach-tooltip-trigger=\\"\\"
            ><svg
              fill=\\"currentColor\\"
              aria-hidden=\\"true\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align: middle\\"
            >
              <g>
                <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
                <path
                  d=\\"M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z\\"
                ></path>
              </g></svg
          ></span>
        </div>
      </button>
      <h2 id=\\"Endre-stiler\\" tabindex=\\"0\\">Endre stiler</h2>
    </div>
  </div>
  <p>Å endre layout, synlighet og design er enda mer vanlig:</p>
  <pre><code>&lt;h1 <strong>onclick</strong>=\\"style='background:yellow'\\"&gt;Min overskrift&lt;/h1&gt;<br></code></pre>
  <p>
    Med denne koden får overskriften gul bakgrunnsfarge. Med stiler kan vi endre
    både synlighet, størrelse, animasjon og annet.
  </p>
  <style data-emotion=\\"css 8s0se0-ContainerDiv\\">
    .css-8s0se0-ContainerDiv {
      position: relative;
    }
    .css-8s0se0-ContainerDiv:hover button {
      cursor: pointer;
      opacity: 0.5;
    }
  </style>
  <div
    data-header-copy-container=\\"true\\"
    data-title=\\"Validering\\"
    class=\\"css-8s0se0-ContainerDiv e18czas70\\"
  >
    <div>
      <style data-emotion=\\"css kgto6a-IconButton\\">
        .css-kgto6a-IconButton {
          position: absolute;
          left: -3em;
          top: 0.1em;
          background: none;
          border: 0;
          z-index: 1;
          -webkit-transition: 0.2s;
          transition: 0.2s;
          opacity: 0;
        }
        .css-kgto6a-IconButton svg {
          width: 30px;
          height: 30px;
        }</style
      ><button data-title=\\"Validering\\" class=\\"css-kgto6a-IconButton e18czas71\\">
        <div
          data-tooltip=\\"true\\"
          data-tooltip-label=\\"Kopier lenke til overskriften\\"
          data-tooltip-from-article-converter=\\"true\\"
        >
          <style data-emotion=\\"css qyvxt5-StyledTooltip\\">
            .css-qyvxt5-StyledTooltip {
              color: #fff;
              z-index: 99999;
              border: 0;
              background: #444;
              border-radius: 4px;
              padding: 6.5px 13px;
              font-family: \\"Source Sans Pro\\", Helvetica, Arial, STKaiti,
                \\"华文楷体\\", KaiTi, SimKai, \\"楷体\\", KaiU, DFKai-SB, \\"標楷體\\",
                SongTi, \\"宋体\\", sans-serif;
              font-size: 14;
              line-height: 1.2;
              font-weight: 400;
              text-align: center;
              white-space: normal;
              max-width: calc(100vw - #{26px});
            }</style
          ><span
            data-tooltip-children=\\"true\\"
            data-state=\\"tooltip-hidden\\"
            data-reach-tooltip-trigger=\\"\\"
            ><svg
              fill=\\"currentColor\\"
              aria-hidden=\\"true\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align: middle\\"
            >
              <g>
                <path d=\\"M0 0h24v24H0z\\" fill=\\"none\\"></path>
                <path
                  d=\\"M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7c-2.76 0-5 2.24-5 5s2.24 5 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1zM8 13h8v-2H8v2zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4c2.76 0 5-2.24 5-5s-2.24-5-5-5z\\"
                ></path>
              </g></svg
          ></span>
        </div>
      </button>
      <h2 id=\\"Validering\\" tabindex=\\"0\\">Validering</h2>
    </div>
  </div>
  <p>
    Mange nettsider har ulike skjema. Det kan være for søk eller innsending av
    forespørsler. Her kan det fort bli feiltasting siden brukeren er fri til å
    skrive inn det han ønsker. Skript vil her komme til nytte både for å veilede
    brukeren og for å sikre at riktig informasjon blir utfylt.
  </p>
  <p>For å sjekke et telefonnummer trengs et skript som sjekker:</p>
  <ul>
    <li>om det kun er skrevet inn tall</li>
    <li>om det er åtte siffer</li>
  </ul>
  <p>
    Skriptet kan da gi en feilmelding om «Feilskrevet telefonnummer» eller en
    mer detaljert feilmelding som «Du har skrevet tegn som ikke er tall».
    Sannsynligvis burde skriptet være så komplekst at det tok høyde for valgfri
    landskode (f.eks. +47).
  </p>
  <p>Vi kan bruke valg-setninger for å teste:</p>
  <pre><code><strong>if</strong> (telefonFelt.value.length != 8) {    /* != betyr ulik */<br>         alert(\\"Du må skrive inn åtte tall.\\");     <br> } </code></pre>
  <p>
    Bruk av <strong>valg</strong> er en av de store styrkene til et
    programmeringsspråk, noe du fort vil erfare. Programmering av
    <strong>gjentakelser</strong> med løkker er en annen. Vi kan sette opp en
    løkke som for eksempel repeterer samme testkode for alle felt i skjemaet
    vårt.
  </p>
</section>
<section>
  <section class=\\"c-related-articles\\">
    <style data-emotion=\\"css 1i9wngt-StyledWrapper\\">
      .css-1i9wngt-StyledWrapper {
        font-weight: 700;
        text-transform: uppercase;
        letter-spacing: 0.05em;
        font-size: 18px;
        font-size: 1rem;
        line-height: 24px;
      }
      @media (min-width: 37.5625em) {
        .css-1i9wngt-StyledWrapper {
          font-size: 20px;
          font-size: 1.1111111111111112rem;
          line-height: 26px;
        }
      }
    </style>
    <h2
      class=\\"
        c-related-articles__component-title
        css-1i9wngt-StyledWrapper
        egsow3l0
      \\"
    >
      Relatert innhold
    </h2>
    <div class=\\"c-related-articles__articles\\">
      <article
        class=\\"
          c-related-articles__item
          c-related-articles__item--external-learning-resources
        \\"
      >
        <h3 class=\\"c-related-articles__title\\" tabindex=\\"0\\">
          <div
            class=\\"
              c-content-type-badge
              c-content-type-badge--external-learning-resources
              c-content-type-badge--small
              c-content-type-badge--background
              c-content-type-badge--border
            \\"
          >
            <svg
              fill=\\"currentColor\\"
              aria-hidden=\\"true\\"
              preserveAspectRatio=\\"xMidYMid meet\\"
              height=\\"1em\\"
              width=\\"1em\\"
              class=\\"c-icon\\"
              viewBox=\\"0 0 24 24\\"
              data-license=\\"Apache License 2.0\\"
              data-source=\\"Material Design\\"
              style=\\"vertical-align: middle\\"
            >
              <g>
                <g
                  stroke=\\"none\\"
                  fill=\\"none\\"
                  stroke-width=\\"1\\"
                  fill-rule=\\"evenodd\\"
                >
                  <g>
                    <g>
                      <polygon points=\\"0 0 24 0 24 24 0 24\\"></polygon>
                      <path
                        d=\\"M3.9,12 C3.9,10.29 5.29,8.9 7,8.9 L11,8.9 L11,7 L7,7 C4.24,7 2,9.24 2,12 C2,14.76 4.24,17 7,17 L11,17 L11,15.1 L7,15.1 C5.29,15.1 3.9,13.71 3.9,12 Z M8,13 L16,13 L16,11 L8,11 L8,13 Z M17,7 L13,7 L13,8.9 L17,8.9 C18.71,8.9 20.1,10.29 20.1,12 C20.1,13.71 18.71,15.1 17,15.1 L13,15.1 L13,17 L17,17 C19.76,17 22,14.76 22,12 C22,9.24 19.76,7 17,7 Z\\"
                        fill=\\"currentColor\\"
                        fill-rule=\\"nonzero\\"
                      ></path>
                    </g>
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <span class=\\"c-related-articles__link-wrapper\\"
            ><a
              href=\\"https://www.codecademy.com/learn/introduction-to-javascript\\"
              class=\\"c-related-articles__link\\"
              target=\\"_blank\\"
              rel=\\"noopener noreferrer\\"
              >Introduction To JavaScript | Codecademy</a
            ></span
          >
        </h3>
        <p class=\\"c-related-articles__description\\"></p>
        <p class=\\"c-related-articles__link-info\\">Nettside hos codecademy.com</p>
      </article>
    </div>
  </section>
</section>
"
`;
